test/DataCollectors/TraceDataCollector.UnitTests/DynamicCoverageDataCollectorImplTests.cs

```diff
@@ -0,0 +1,374 @@
+        }
+
+        #region Initialize Tests
+        [TestMethod]
+        public void InitializeShouldCreateDefaultCodeCoverageSettingsIfConfigElementIsNull()
+        {
+            this.collectorImpl.Initialize(null, this.dataCollectionSinkMock.Object, this.dataCollectionLoggerMock.Object);
+
+            Assert.AreEqual(DynamicCoverageDataCollectorImplTests.DefaultConfigFileName, Path.GetFileName(aConfigFileName));
+            StringAssert.Contains(aConfigFileName, Path.GetTempPath());
+            Assert.AreEqual(
+                DefaultCodeCoverageConfig.Replace(Environment.NewLine, string.Empty).Replace(" ", String.Empty)
+                , aConfig.InnerXml.Replace(Environment.NewLine, string.Empty).Replace(" ", String.Empty));
+        }
+
+        [TestMethod]
+        public void InitializeShouldInitializeVanguardWithRightCoverageSettings()
+        {
+            var expectedContent = "CoverageSettingsContent";
+            XmlElement configElement =
+                DynamicCoverageDataCollectorImplTests.CreateXmlElement($@"<Configuration><CodeCoverage>{expectedContent}</CodeCoverage></Configuration>");
+
+            this.collectorImpl.Initialize(configElement, this.dataCollectionSinkMock.Object, this.dataCollectionLoggerMock.Object);
+
+            Assert.AreEqual(DynamicCoverageDataCollectorImplTests.DefaultConfigFileName, Path.GetFileName(aConfigFileName));
+            Assert.AreEqual(expectedContent, aConfig.InnerXml);
+        }
+
+        [TestMethod]
+        public void InitializeShouldRegisterForSendFileCompleteEvent()
+        {
+            this.dataCollectionSinkMock.Raise(s => s.SendFileCompleted += null, new AsyncCompletedEventArgs(null, false, null));
+            this.directoryHelperMock.Verify(d => d.Delete(this.atempDirectory, true));
+        }
+
+        [TestMethod]
+        public void InitializeShouldCreateTempDirectoryForSession()
+        {
+            this.directoryHelperMock.Verify(d => d.CreateDirectory(this.atempDirectory));
+        }
+        #endregion
+
+        #region Dispose Tests
+        [TestMethod]
+        public void DisposeShouldStopVanguard()
+        {
+            this.collectorImpl.Dispose();
+            this.vangurdMock.Verify(v =>v.Stop());
+        }
+
+        [TestMethod]
+        public void DisposeShouldDisposeVanguard()
+        {
+            this.collectorImpl.Dispose();
+            this.vangurdMock.Verify(v => v.Stop());
+        }
+
+        [TestMethod]
+        public void DisposeShouldDeleteTempDirectory()
+        {
+            this.collectorImpl.Dispose();
+            this.directoryHelperMock.Verify(d => d.Delete(this.atempDirectory, true));
+        }
+
+        [TestMethod]
+        public void DisposeShouldUnregisterFileCompleteEvent()
+        {
+            this.collectorImpl.Dispose();
+            this.dataCollectionSinkMock.Raise(s => s.SendFileCompleted += null, new AsyncCompletedEventArgs(null, false, null));
```

Comment:
How are we checking if event is unregistered

Reply:
If not unregistered, directoryHelperMock method calls twice.